File: 70-climbing-stairs/70-climbing-stairs.java
class Solution {
    public int climbStairs(int n) {
        return totalWays(0,n,new HashMap<Integer,Integer>());
            
    }
        public int totalWays(int cs,int ts,HashMap<Integer,Integer> memo)
        {
                if(cs==ts)
                {
                        return 1;

                }
                if(cs>ts)
                {
                        return 0;
                }
                int currentKey=cs;
                if(memo.containsKey(currentKey))
                {
                        return memo.get(currentKey);
                }
                int onestep=totalWays(cs+1,ts,memo);
                int twostep=totalWays(cs+2,ts,memo);
                memo.put(cs,onestep+twostep);
                return onestep+twostep;
        }
}

File: 70-climbing-stairs/70-climbing-stairs.cpp
class Solution {
public:
    int climbStairs(int n) {
        vector<int> dp(n+1,-1);
        return findways(n,dp);
    }
    int findways(int n,vector<int>&dp)
    {
        if(n<=1)
        {
            return 1;
        }
        if(dp[n]!=-1)
        {
            return dp[n];
        }
        return dp[n]=findways(n-1,dp)+findways(n-2,dp);
    }
};